File: cli/vaux/crypto.py
"""Private-room cryptography. Mirrors web/lib/crypto.ts byte-for-byte —
any change here must be reflected there. See PRIVATE_ROOMS_SPEC.md for
the full spec, parameter table, and threat model."""

from __future__ import annotations

import base64
import hashlib
import os
import re
from dataclasses import dataclass
from urllib.parse import urlparse

from nacl import bindings as _nb
from nacl.secret import SecretBox
from nacl.utils import random as _nacl_random

PASSWORD_BYTES = 16
PASSWORD_REGEX = re.compile(r"^[A-Za-z0-9_-]{22}$")
ROOM_ID_BYTES = 16

ARGON2_OPSLIMIT = 2
ARGON2_MEMLIMIT_BYTES = 67_108_864
ARGON2_OUTLEN = 32
# libsodium ALG_ARGON2ID13. pynacl exposes no named constant for this value.
ARGON2_ALG = 2

KDF_SUBKEY_LEN = 32
KDF_CTX_RID = b"vaux/rid"
KDF_CTX_ATH = b"vaux/ath"
KDF_CTX_CHT = b"vaux/cht"
KDF_SUBKEY_RID = 1
KDF_SUBKEY_ATH = 2
KDF_SUBKEY_CHT = 3

# SHA-256("vaux/private-room/v1")[:16]. Pinned. Bumping requires v2 marker.
VAUX_PRIVATE_SALT_HEX = "f4c839505e351f0d2d8733459e2db801"
VAUX_PRIVATE_SALT = bytes.fromhex(VAUX_PRIVATE_SALT_HEX)


# ---------------------------------------------------------------------------
# Password helpers
# ---------------------------------------------------------------------------

def generate_password() -> str:
    """22-char base64url-unpadded password (~128 bits entropy)."""
    raw = os.urandom(PASSWORD_BYTES)
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def is_well_formed_password(password: str) -> bool:
    return bool(PASSWORD_REGEX.match(password or ""))


def parse_invite(text: str) -> str | None:
    """Extract a 22-char base64url password from a raw string or invite URL.

    Accepts:
      - bare password: "AAAAAAAAAAAAAAAAAAAAAA"
      - URL with fragment: "http://host/#<password>"
      - URL with fragment + leading '#': "#<password>"
    Returns None if no well-formed password is found.
    """
    if not text:
        return None
    s = text.strip()

    # URL-with-fragment form
    if "://" in s or s.startswith("/") or s.startswith("#"):
        if s.startswith("#"):
            candidate = s[1:]
        else:
            try:
                parsed = urlparse(s)
            except ValueError:
                return None
            candidate = parsed.fragment or ""
        candidate = candidate.split("?")[0].split("&")[0]
        return candidate if is_well_formed_password(candidate) else None

    # Raw password form
    return s if is_well_formed_password(s) else None


# ---------------------------------------------------------------------------
# Key derivation
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class RoomMaterial:
    room_id: str
    auth_proof: bytes
    chat_key: bytes


def _kdf(master_key: bytes, subkey_id: int, ctx: bytes) -> bytes:
    """libsodium crypto_kdf_derive_from_key with a hardcoded blake2b fallback
    in case the pynacl build doesn't surface the binding."""
    assert len(master_key) == 32
    assert len(ctx) == 8
    fn = getattr(_nb, "crypto_kdf_derive_from_key", None)
    if fn is not None:
        return fn(KDF_SUBKEY_LEN, subkey_id, ctx, master_key)
    # Fallback matches libsodium's blake2b-based KDF construction:
    # blake2b(personal=ctx, salt=subkey_id_le, key=master_key, msg=b"")
    return hashlib.blake2b(
        b"",
        digest_size=KDF_SUBKEY_LEN,
        key=master_key,
        salt=subkey_id.to_bytes(8, "little"),
        person=ctx,
    ).digest()


def derive_room_material(password: str) -> RoomMaterial:
    """Argon2id (~250 ms) + KDF over a fixed application salt. Cache the
    result for the session — never re-derive per chat message."""
    pw_bytes = password.encode("utf-8")
    # bindings.crypto_pwhash is a submodule; the actual fn we need is
    # crypto_pwhash_alg (raw KDF, alg-selectable). The libsodium constant
    # for ALG_ARGON2ID13 is 2 — pinned in ARGON2_ALG.
    master_key = _nb.crypto_pwhash_alg(
        ARGON2_OUTLEN,
        pw_bytes,
        VAUX_PRIVATE_SALT,
        ARGON2_OPSLIMIT,
        ARGON2_MEMLIMIT_BYTES,
        ARGON2_ALG,
    )

    rid_full = _kdf(master_key, KDF_SUBKEY_RID, KDF_CTX_RID)
    room_id = (
        base64.urlsafe_b64encode(rid_full[:ROOM_ID_BYTES])
        .rstrip(b"=")
        .decode("ascii")
    )
    auth_proof = _kdf(master_key, KDF_SUBKEY_ATH, KDF_CTX_ATH)
    chat_key = _kdf(master_key, KDF_SUBKEY_CHT, KDF_CTX_CHT)

    return RoomMaterial(room_id=room_id, auth_proof=auth_proof, chat_key=chat_key)


# ---------------------------------------------------------------------------
# Encoding helpers — base64 (NOT base64url) on the wire to match web client.
# ---------------------------------------------------------------------------

def auth_proof_to_b64(auth_proof: bytes) -> str:
    return base64.b64encode(auth_proof).decode("ascii")


def bytes_to_b64(b: bytes) -> str:
    return base64.b64encode(b).decode("ascii")


def b64_to_bytes(b64: str) -> bytes:
    return base64.b64decode(b64, validate=False)


# ---------------------------------------------------------------------------
# SecretBox (XSalsa20-Poly1305) — chat + username payloads
# ---------------------------------------------------------------------------

def encrypt_chat(chat_key: bytes, text: str) -> dict[str, str]:
    """Returns {"ct": <b64>, "nonce": <b64>}. Wire format identical to JS.
    `EncryptedMessage.bytes` would prepend the nonce to the ciphertext —
    we use `.ciphertext` so the format matches sodium.crypto_secretbox_easy."""
    box = SecretBox(chat_key)
    nonce = _nacl_random(SecretBox.NONCE_SIZE)
    encrypted = box.encrypt(text.encode("utf-8"), nonce)
    return {
        "ct": bytes_to_b64(encrypted.ciphertext),
        "nonce": bytes_to_b64(nonce),
    }


def decrypt_chat(chat_key: bytes, ct_b64: str, nonce_b64: str) -> str | None:
    """Returns plaintext, or None on tamper / wrong key / malformed input.
    Callers must drop None silently — never bubble decrypt errors to the UI."""
    try:
        box = SecretBox(chat_key)
        ct = b64_to_bytes(ct_b64)
        nonce = b64_to_bytes(nonce_b64)
        return box.decrypt(ct, nonce).decode("utf-8")
    except Exception:
        return None

File: web/lib/crypto.ts
// Private-room cryptography. Mirrors cli/vaux/crypto.py byte-for-byte —
// any change here must be reflected there. See PRIVATE_ROOMS_SPEC.md for
// the full spec, parameter table, and threat model.

// `-sumo` build includes Argon2id (crypto_pwhash). The minimal
// libsodium-wrappers build omits it.
import sodium from "libsodium-wrappers-sumo";

const PASSWORD_BYTES = 16;
const PASSWORD_REGEX = /^[A-Za-z0-9_-]{22}$/;
const ROOM_ID_BYTES = 16;
const ARGON2_OPSLIMIT = 2;
const ARGON2_MEMLIMIT_BYTES = 67_108_864;
const ARGON2_OUTLEN = 32;
const KDF_SUBKEY_LEN = 32;
const KDF_CTX_RID = "vaux/rid";
const KDF_CTX_ATH = "vaux/ath";
const KDF_CTX_CHT = "vaux/cht";
const KDF_SUBKEY_RID = 1;
const KDF_SUBKEY_ATH = 2;
const KDF_SUBKEY_CHT = 3;

// SHA-256("vaux/private-room/v1")[:16]. Pinned. See PRIVATE_ROOMS_SPEC.md
// for derivation. Bumping requires a v2 protocol marker.
const VAUX_PRIVATE_SALT_HEX = "f4c839505e351f0d2d8733459e2db801";

let saltCache: Uint8Array | null = null;
let readyPromise: Promise<void> | null = null;

async function ensureReady(): Promise<void> {
  if (!readyPromise) readyPromise = sodium.ready;
  await readyPromise;
}

function getSalt(): Uint8Array {
  if (!saltCache) saltCache = sodium.from_hex(VAUX_PRIVATE_SALT_HEX);
  return saltCache;
}

export type RoomMaterial = {
  roomId: string;
  authProof: Uint8Array;
  chatKey: Uint8Array;
};

/** New random 22-char base64url password. ~128 bits entropy. */
export async function generatePassword(): Promise<string> {
  await ensureReady();
  const bytes = sodium.randombytes_buf(PASSWORD_BYTES);
  return sodium.to_base64(bytes, sodium.base64_variants.URLSAFE_NO_PADDING);
}

export function isWellFormedPassword(password: string): boolean {
  return PASSWORD_REGEX.test(password);
}

/**
 * Derive room material from a password. Argon2id is intentionally slow
 * (~250 ms target) — call once per session and cache the result.
 */
export async function deriveRoomMaterial(
  password: string,
): Promise<RoomMaterial> {
  await ensureReady();

  const passwordBytes = new TextEncoder().encode(password);
  const masterKey = sodium.crypto_pwhash(
    ARGON2_OUTLEN,
    passwordBytes,
    getSalt(),
    ARGON2_OPSLIMIT,
    ARGON2_MEMLIMIT_BYTES,
    sodium.crypto_pwhash_ALG_ARGON2ID13,
  );

  const roomIdBytes = sodium.crypto_kdf_derive_from_key(
    KDF_SUBKEY_LEN,
    KDF_SUBKEY_RID,
    KDF_CTX_RID,
    masterKey,
  );
  const roomId = sodium.to_base64(
    roomIdBytes.subarray(0, ROOM_ID_BYTES),
    sodium.base64_variants.URLSAFE_NO_PADDING,
  );

  const authProof = sodium.crypto_kdf_derive_from_key(
    KDF_SUBKEY_LEN,
    KDF_SUBKEY_ATH,
    KDF_CTX_ATH,
    masterKey,
  );

  const chatKey = sodium.crypto_kdf_derive_from_key(
    KDF_SUBKEY_LEN,
    KDF_SUBKEY_CHT,
    KDF_CTX_CHT,
    masterKey,
  );

  return { roomId, authProof, chatKey };
}

export function authProofToB64(authProof: Uint8Array): string {
  return sodium.to_base64(authProof, sodium.base64_variants.ORIGINAL);
}

export function bytesToB64(bytes: Uint8Array): string {
  return sodium.to_base64(bytes, sodium.base64_variants.ORIGINAL);
}

export function b64ToBytes(b64: string): Uint8Array {
  return sodium.from_base64(b64, sodium.base64_variants.ORIGINAL);
}

export type EncryptedChat = { ct: string; nonce: string };

export async function encryptChat(
  chatKey: Uint8Array,
  text: string,
): Promise<EncryptedChat> {
  await ensureReady();
  const nonce = sodium.randombytes_buf(sodium.crypto_secretbox_NONCEBYTES);
  const plaintext = new TextEncoder().encode(text);
  const ct = sodium.crypto_secretbox_easy(plaintext, nonce, chatKey);
  return {
    ct: sodium.to_base64(ct, sodium.base64_variants.ORIGINAL),
    nonce: sodium.to_base64(nonce, sodium.base64_variants.ORIGINAL),
  };
}

/**
 * Decrypt chat ciphertext. Returns null on auth failure (tampered ct,
 * wrong key, malformed input). Callers must treat null as "drop the
 * message" — never bubble decrypt errors up to the UI.
 */
export async function decryptChat(
  chatKey: Uint8Array,
  ctB64: string,
  nonceB64: string,
): Promise<string | null> {
  await ensureReady();
  try {
    const ct = sodium.from_base64(ctB64, sodium.base64_variants.ORIGINAL);
    const nonce = sodium.from_base64(
      nonceB64,
      sodium.base64_variants.ORIGINAL,
    );
    const plaintext = sodium.crypto_secretbox_open_easy(ct, nonce, chatKey);
    return new TextDecoder().decode(plaintext);
  } catch {
    return null;
  }
}
